File: 9. Delegati in lambde/delegati_in_lambde.cs
// Delegat:
// Delegate v C# je tip, ki omogoča shranjevanje reference na metodo s specifičnim podpisom. 
// Uporablja se za podajanje metod kot argumentov ali za obvladovanje dogodkov.

delegate int Operacija(int a, int b);
Operacija seštej = (a, b) => a + b;
int rezultat = seštej(5, 3); // Rezultat bo 8

//////////////////////////////////////////////////////////////////////////////////////

// Lambda izraz:
// Lambda izraz je anonimna funkcija, ki jo lahko shranimo v spremenljivko ali jo podamo kot argument.
// Lambda izraz je lahko bodisi izraz bodisi stavek bloka.

// Izraz:
(int a, int b) => a + b;

// Stavek bloka:
(int a, int b) => { return a + b; };

// Izpiši vsak element v seznamu:
List<int> list = new List<int> { 1, 2, 3, 4, 5 };
list.ForEach(item => Console.WriteLine(item));
// Rezultat: 1 2 3 4 5

// Izpiši vsak element v seznamu, ki je večji od 3:
List<int> list = new List<int> { 1, 2, 3, 4, 5 };
list.Where(item => item > 3).ToList().ForEach(item => Console.WriteLine(item));
// Rezultat: 4 5

//////////////////////////////////////////////////////////////////////////////////////

// LINQ:
// Language Integrated Query (LINQ) je tehnologija vgrajena v C#, 
// ki omogoča poizvedovanje nad podatkovnimi viri.

// Izpiši vsak element v seznamu, ki je večji od 3:
List<int> list = new List<int> { 1, 2, 3, 4, 5 };
var query = from item in list
            where item > 3
            select item;

foreach (var item in query)
{
    Console.WriteLine(item);
}
// Rezultat: 4 5

//////////////////////////////////////////////////////////////////////////////////////

File: 9. Delegati in lambde/delegati_in_lambde.java
// Delegat:
// Delegat omogoča shranjevanje reference na metodo s specifičnim podpisom. 
// Uporablja se za podajanje metod kot argumentov ali za obvladovanje dogodkov.

interface Calculator {
    int calculate(int a, int b);
}

class AddCalculator implements Calculator {
    @Override
    public int calculate(int a, int b) {
        return a + b;
    }
}

public class Main {
    public static void main(String[] args) {
        Calculator calculator = new AddCalculator();
        int result = calculator.calculate(5, 3);
        System.out.println("Rezultat: " + result);
    }
}

//////////////////////////////////////////////////////////////////////////////////////

// Lambda izraz:
// Lambda izraz je anonimna funkcija, ki jo lahko shranimo v spremenljivko ali jo podamo kot argument.
// Lambda izraz je lahko bodisi izraz bodisi stavek bloka.

ArrayList<Integer> numbers = new ArrayList<Integer>();
numbers.add(5);
numbers.add(9);
numbers.add(8);
numbers.add(1);
numbers.forEach( (n) -> { System.out.println(n); } );
// Rezultat: 5 9 8 1

// Lambda izraz je lahko shranjen v spremenljivko.
Calculator calculator = (a, b) -> a + b;
int result = calculator.calculate(5, 3);
System.out.println("Rezultat: " + result);
// Rezultat: 8

//////////////////////////////////////////////////////////////////////////////////////